File: src/services/digest.js
/**
 * Morning digest service.
 * Sends each team member their open tasks at the team's configured hour.
 *
 * Run via node-cron every minute — checks if any team's digest should fire now.
 */

const cron = require('node-cron');
const db = require('../db');
const sms = require('./sms');
const { formatDueDate, isOverdue, isToday } = require('../utils/dates');

/**
 * Send the morning digest for a single team.
 */
async function sendTeamDigest(team) {
  const users = await db.getTeamUsers(team.id);

  for (const user of users) {
    const tasks = await db.getOpenTasksForUser(user.id);
    if (tasks.length === 0) continue;

    const overdueItems = tasks.filter(t => t.due_date && isOverdue(t.due_date));
    const todayItems = tasks.filter(t => t.due_date && isToday(t.due_date));
    const upcoming = tasks.filter(t => !t.due_date || (!isOverdue(t.due_date) && !isToday(t.due_date)));

    const lines = [];
    lines.push(`Good morning, ${user.name}! Here's your day:`);
    lines.push('');

    if (overdueItems.length) {
      lines.push('⚠ Overdue:');
      overdueItems.forEach(t => lines.push(`  • ${t.title} (${formatDueDate(t.due_date)})`));
      lines.push('');
    }

    if (todayItems.length) {
      lines.push('📌 Due today:');
      todayItems.forEach(t => lines.push(`  • ${t.title}`));
      lines.push('');
    }

    if (upcoming.length) {
      lines.push('Up next:');
      upcoming.slice(0, 5).forEach(t => {
        const due = t.due_date ? ` (${formatDueDate(t.due_date)})` : '';
        lines.push(`  • ${t.title}${due}`);
      });
    }

    const body = lines.join('\n');

    try {
      await sms.send({ to: user.phone_number, body, teamId: team.id, userId: user.id });
      console.log(`[Digest] Sent to ${user.name} (${user.phone_number})`);
    } catch (err) {
      console.error(`[Digest] Failed for ${user.name}:`, err.message);
    }
  }
}

/**
 * Check all teams and fire digest if it's their configured hour (local time).
 */
async function checkAndSendDigests() {
  const teams = await db.getAllTeams();
  const nowUTC = new Date();

  for (const team of teams) {
    if (!team.setup_complete) continue;

    // Convert UTC now to team's local hour
    const localHour = getLocalHour(nowUTC, team.timezone);

    if (localHour === team.digest_hour) {
      console.log(`[Digest] Sending for team ${team.name} (local hour: ${localHour})`);
      await sendTeamDigest(team);
    }
  }
}

/**
 * Get the current hour in a given IANA timezone.
 */
function getLocalHour(date, timezone) {
  try {
    const formatter = new Intl.DateTimeFormat('en-US', {
      timeZone: timezone,
      hour: 'numeric',
      hour12: false,
    });
    const parts = formatter.formatToParts(date);
    const hourPart = parts.find(p => p.type === 'hour');
    return parseInt(hourPart?.value || '0', 10);
  } catch {
    // Fallback to UTC
    return date.getUTCHours();
  }
}

/**
 * Start the cron job — runs once per minute.
 * Only sends if current minute is :00 (on the hour).
 */
function startDigestCron() {
  // Run at the top of every hour
  cron.schedule('0 * * * *', async () => {
    console.log('[Digest] Hourly check running...');
    try {
      await checkAndSendDigests();
    } catch (err) {
      console.error('[Digest] Error:', err.message);
    }
  });

  console.log('[Digest] Cron started — will check hourly for teams due a digest.');
}

module.exports = { startDigestCron, sendTeamDigest, checkAndSendDigests };

File: core/services/digest.py
import logging
from datetime import datetime, timezone
from zoneinfo import ZoneInfo

from apscheduler.schedulers.background import BackgroundScheduler

from core.utils import format_due_date, is_overdue, is_today

logger = logging.getLogger(__name__)


def send_team_digest(team):
    """Send each team member their morning task summary."""
    from core.models import User
    from core.services import sms_service

    for user in User.objects.filter(team=team):
        tasks = list(user.owned_tasks.filter(status='open').order_by('due_date'))
        if not tasks:
            continue

        due_strs = [str(t.due_date) if t.due_date else None for t in tasks]

        overdue = [t for t, d in zip(tasks, due_strs) if d and is_overdue(d)]
        today = [t for t, d in zip(tasks, due_strs) if d and is_today(d)]
        upcoming = [t for t, d in zip(tasks, due_strs) if not d or (not is_overdue(d) and not is_today(d))]

        lines = [f"Good morning, {user.name}! Here's your day:", '']

        if overdue:
            lines.append('\u26a0 Overdue:')
            for t in overdue:
                lines.append(f'  \u2022 {t.title} ({format_due_date(str(t.due_date))})')
            lines.append('')

        if today:
            lines.append('\U0001f4cc Due today:')
            for t in today:
                lines.append(f'  \u2022 {t.title}')
            lines.append('')

        if upcoming:
            lines.append('Up next:')
            for t in upcoming[:5]:
                due = f' ({format_due_date(str(t.due_date))})' if t.due_date else ''
                lines.append(f'  \u2022 {t.title}{due}')

        body = '\n'.join(lines)

        try:
            sms_service.send(to=user.phone_number, body=body, team_id=str(team.id), user_id=str(user.id))
            logger.info('[Digest] Sent to %s (%s)', user.name, user.phone_number)
        except Exception as e:
            logger.error('[Digest] Failed for %s: %s', user.name, e)


def check_and_send_digests():
    """Run at the top of each hour; fire digests for any team whose local hour matches."""
    from django.db import close_old_connections
    close_old_connections()

    from core.models import Team

    now_utc = datetime.now(timezone.utc)
    for team in Team.objects.filter(setup_complete=True):
        local_hour = _get_local_hour(now_utc, team.timezone)
        if local_hour == team.digest_hour:
            logger.info('[Digest] Sending for team %s (local hour: %d)', team.name, local_hour)
            try:
                send_team_digest(team)
            except Exception as e:
                logger.error('[Digest] Error for team %s: %s', team.name, e)


def _get_local_hour(dt, timezone_str):
    try:
        return dt.astimezone(ZoneInfo(timezone_str)).hour
    except Exception:
        return dt.utctimetuple().tm_hour


def start_digest_cron():
    scheduler = BackgroundScheduler()
    scheduler.add_job(check_and_send_digests, 'cron', minute=0)
    scheduler.start()
    logger.info('[Digest] APScheduler started \u2014 will check hourly for teams due a digest.')
